File: 2560. House Robber IV/main.cpp
#include <algorithm>
#include <limits>
#include <vector>

using std::max;
using std::min;
using std::vector;

class Solution {
public:
    int minCapability(vector<int> &nums, int k) {
        int left = std::numeric_limits<int>::max();
        int right = std::numeric_limits<int>::min();
        int size = nums.size();

        for (const auto &num : nums) {
            if (num < left) {
                left = num;
            }
            if (num > right) {
                right = num;
            }
        }

        while (left < right) {
            int mid = (left + right) / 2;
            int count = 0;
            int index = 0;
            while (index < size) {
                if (nums[index] <= mid) {
                    count += 1;
                    index += 2;
                } else {
                    index += 1;
                }
            }

            if (count >= k) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }
};

class Solution {
public:
    int minCapability(vector<int> &nums, int k) {
        int left = std::numeric_limits<int>::max();
        int right = std::numeric_limits<int>::min();
        int size = nums.size();

        for (const auto &num : nums) {
            if (num < left) {
                left = num;
            }
            if (num > right) {
                right = num;
            }
        }

        while (left < right) {
            int mid = (left + right) / 2;
            if (canRob(nums, mid, k)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

private:
    bool canRob(vector<int> &nums, int max, int k) {
        int count = 0;
        int index = 0;
        while (index < nums.size()) {
            if (nums[index] <= max) {
                count += 1;
                index += 2;
            } else {
                index += 1;
            }
        }

        return count >= k;
    }
};

File: 2560. House Robber IV/main.py
from typing import List


class Solution:
    def minCapability(self, nums: List[int], k: int) -> int:
        left = min(nums)
        right = max(nums)

        while left < right:
            mid = (left + right) // 2
            count = 0
            index = 0
            while index < len(nums):
                if nums[index] <= mid:
                    count += 1
                    index += 2
                else:
                    index += 1

            if count >= k:
                right = mid
            else:
                left = mid + 1

        return left


from typing import List


class Solution:
    def minCapability(self, nums: List[int], k: int) -> int:
        left, right = min(nums), max(nums)

        while left < right:
            mid = (left + right) // 2
            if self.canRob(nums, mid, k):
                right = mid
            else:
                left = mid + 1

        return left

    def canRob(self, nums: List[int], cap: int, k: int) -> bool:
        count = 0
        index = 0
        while index < len(nums):
            if nums[index] <= cap:
                count += 1
                index += 2
            else:
                index += 1

        return count >= k
